File: implementations/java-springboot/examples/src/test/java/com/example/accountservice/account/application/command/AnalyzeMonthlySpendingServiceTest.java
package com.example.accountservice.account.application.command;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.example.accountservice.account.domain.Account;
import com.example.accountservice.account.domain.AccountFindQuery;
import com.example.accountservice.account.domain.AccountRepository;
import com.example.accountservice.account.domain.AccountStatus;
import com.example.accountservice.account.domain.AccountsWithCount;
import com.example.accountservice.account.domain.SpendingAnalysis;
import com.example.accountservice.account.domain.SpendingAnalysisRepository;
import com.example.accountservice.account.domain.SpendingTrend;
import com.example.accountservice.account.domain.TransactionSummary;
import com.example.accountservice.account.domain.TransactionSummaryQuery;
import com.example.accountservice.account.domain.TransactionType;
import java.time.LocalDateTime;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class AnalyzeMonthlySpendingServiceTest {

    @Mock private AccountRepository accountRepository;
    @Mock private SpendingAnalysisRepository spendingAnalysisRepository;

    private AnalyzeMonthlySpendingService service;

    private static final LocalDateTime MONTH_START = LocalDateTime.of(2026, 7, 1, 0, 0);
    private static final LocalDateTime MONTH_END = LocalDateTime.of(2026, 8, 1, 0, 0);
    private static final LocalDateTime PREVIOUS_MONTH_START = LocalDateTime.of(2026, 6, 1, 0, 0);
    private static final LocalDateTime PREVIOUS_MONTH_END = MONTH_START;

    @BeforeEach
    void setUp() {
        service = new AnalyzeMonthlySpendingService(accountRepository, spendingAnalysisRepository);
    }

    private Account activeAccount() {
        return Account.create("owner-1", "owner-1@example.com", "KRW");
    }

    private AnalyzeMonthlySpendingCommand command() {
        return new AnalyzeMonthlySpendingCommand(
                "2026-07", MONTH_START, MONTH_END, PREVIOUS_MONTH_START, PREVIOUS_MONTH_END);
    }

    private void stubAccounts(Account account) {
        when(accountRepository.findAccounts(
                        new AccountFindQuery(
                                0, 100, null, null, List.of(AccountStatus.ACTIVE.name()))))
                .thenReturn(new AccountsWithCount(List.of(account), 1));
    }

    @Test
    void summarizes_current_and_previous_month_with_the_right_date_ranges_and_saves_the_analysis() {
        Account account = activeAccount();
        stubAccounts(account);
        when(spendingAnalysisRepository.hasAnalysis(account.getAccountId(), "2026-07"))
                .thenReturn(false);
        when(accountRepository.summarizeTransactions(
                        new TransactionSummaryQuery(
                                account.getAccountId(),
                                List.of(TransactionType.WITHDRAWAL),
                                MONTH_START,
                                MONTH_END)))
                .thenReturn(new TransactionSummary(2, 50000));
        when(accountRepository.summarizeTransactions(
                        new TransactionSummaryQuery(
                                account.getAccountId(),
                                List.of(TransactionType.WITHDRAWAL),
                                PREVIOUS_MONTH_START,
                                PREVIOUS_MONTH_END)))
                .thenReturn(new TransactionSummary(0, 0));

        int analyzedCount = service.analyze(command());

        assertThat(analyzedCount).isEqualTo(1);
        ArgumentCaptor<SpendingAnalysis> captor = ArgumentCaptor.forClass(SpendingAnalysis.class);
        verify(spendingAnalysisRepository, times(1)).saveAnalysis(captor.capture());
        SpendingAnalysis saved = captor.getValue();
        assertThat(saved.getAccountId()).isEqualTo(account.getAccountId());
        assertThat(saved.getAnalysisMonth()).isEqualTo("2026-07");
        assertThat(saved.getTotalAmount()).isEqualTo(50000);
        assertThat(saved.getTransactionCount()).isEqualTo(2);
        assertThat(saved.getAverageAmount()).isEqualTo(25000);
        // No prior-month withdrawal history exists, so the comparison baseline is 0 — the
        // %-change is capped at 100 and the trend is INCREASING (see SpendingAnalysis.create).
        assertThat(saved.getChangeFromPreviousMonth()).isEqualTo(100);
        assertThat(saved.getTrend()).isEqualTo(SpendingTrend.INCREASING);
    }

    @Test
    void skips_an_account_that_has_already_been_analyzed_this_month() {
        Account account = activeAccount();
        stubAccounts(account);
        when(spendingAnalysisRepository.hasAnalysis(account.getAccountId(), "2026-07"))
                .thenReturn(true);

        int analyzedCount = service.analyze(command());

        assertThat(analyzedCount).isEqualTo(0);
        verify(accountRepository, never()).summarizeTransactions(any());
        verify(spendingAnalysisRepository, never()).saveAnalysis(any());
    }

    @Test
    void does_nothing_when_there_are_no_active_accounts() {
        when(accountRepository.findAccounts(
                        new AccountFindQuery(
                                0, 100, null, null, List.of(AccountStatus.ACTIVE.name()))))
                .thenReturn(new AccountsWithCount(List.of(), 0));

        int analyzedCount = service.analyze(command());

        assertThat(analyzedCount).isEqualTo(0);
        verify(spendingAnalysisRepository, never()).saveAnalysis(any());
    }
}

File: implementations/kotlin-springboot/examples/src/test/kotlin/com/example/accountservice/account/application/command/AnalyzeMonthlySpendingServiceTest.kt
package com.example.accountservice.account.application.command

import com.example.accountservice.account.domain.Account
import com.example.accountservice.account.domain.AccountFindQuery
import com.example.accountservice.account.domain.AccountRepository
import com.example.accountservice.account.domain.AccountStatus
import com.example.accountservice.account.domain.SpendingAnalysisRepository
import com.example.accountservice.account.domain.TransactionSummary
import com.example.accountservice.account.domain.TransactionSummaryQuery
import com.example.accountservice.account.domain.TransactionType
import io.mockk.every
import io.mockk.mockk
import io.mockk.verify
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

class AnalyzeMonthlySpendingServiceTest {
    private val accountRepository = mockk<AccountRepository>()
    private val spendingAnalysisRepository = mockk<SpendingAnalysisRepository>(relaxed = true)
    private val service = AnalyzeMonthlySpendingService(accountRepository, spendingAnalysisRepository)

    private val account = Account.create("owner-1", "KRW", "owner-1@example.com")

    private val analysisMonth = "2026-07"
    private val monthStart = LocalDateTime.parse("2026-07-01T00:00:00")
    private val monthEnd = LocalDateTime.parse("2026-08-01T00:00:00")
    private val previousMonthStart = LocalDateTime.parse("2026-06-01T00:00:00")
    private val previousMonthEnd = LocalDateTime.parse("2026-07-01T00:00:00")

    @Test
    fun `when an account has not been analyzed yet then summarizes both months and saves the analysis`() {
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 100, status = listOf(AccountStatus.ACTIVE.name)))
        } returns (listOf(account) to 1L)
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 1, take = 100, status = listOf(AccountStatus.ACTIVE.name)))
        } returns (emptyList<Account>() to 1L)
        every { spendingAnalysisRepository.hasAnalysis(account.accountId, analysisMonth) } returns false
        every {
            accountRepository.summarizeTransactions(
                TransactionSummaryQuery(account.accountId, listOf(TransactionType.WITHDRAWAL), monthStart, monthEnd),
            )
        } returns TransactionSummary(count = 2, totalAmount = 15000)
        every {
            accountRepository.summarizeTransactions(
                TransactionSummaryQuery(account.accountId, listOf(TransactionType.WITHDRAWAL), previousMonthStart, previousMonthEnd),
            )
        } returns TransactionSummary(count = 1, totalAmount = 10000)

        val analyzedCount = service.analyzeMonthlySpending(analysisMonth, monthStart, monthEnd, previousMonthStart, previousMonthEnd)

        verify(exactly = 1) {
            accountRepository.summarizeTransactions(
                TransactionSummaryQuery(account.accountId, listOf(TransactionType.WITHDRAWAL), monthStart, monthEnd),
            )
        }
        verify(exactly = 1) {
            accountRepository.summarizeTransactions(
                TransactionSummaryQuery(account.accountId, listOf(TransactionType.WITHDRAWAL), previousMonthStart, previousMonthEnd),
            )
        }
        verify(exactly = 1) {
            spendingAnalysisRepository.saveAnalysis(
                match {
                    it.accountId == account.accountId &&
                        it.analysisMonth == analysisMonth &&
                        it.totalAmount == 15000L &&
                        it.transactionCount == 2L &&
                        it.trend.name == "INCREASING"
                },
            )
        }
        assertThat(analyzedCount).isEqualTo(1)
    }

    @Test
    fun `when an account was already analyzed this month then skips it`() {
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 100, status = listOf(AccountStatus.ACTIVE.name)))
        } returns (listOf(account) to 1L)
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 1, take = 100, status = listOf(AccountStatus.ACTIVE.name)))
        } returns (emptyList<Account>() to 1L)
        every { spendingAnalysisRepository.hasAnalysis(account.accountId, analysisMonth) } returns true

        val analyzedCount = service.analyzeMonthlySpending(analysisMonth, monthStart, monthEnd, previousMonthStart, previousMonthEnd)

        verify(exactly = 0) { accountRepository.summarizeTransactions(any()) }
        verify(exactly = 0) { spendingAnalysisRepository.saveAnalysis(any()) }
        assertThat(analyzedCount).isEqualTo(0)
    }

    @Test
    fun `when there are no active accounts then returns 0 without summarizing`() {
        every {
            accountRepository.findAccounts(AccountFindQuery(page = 0, take = 100, status = listOf(AccountStatus.ACTIVE.name)))
        } returns (emptyList<Account>() to 0L)

        val analyzedCount = service.analyzeMonthlySpending(analysisMonth, monthStart, monthEnd, previousMonthStart, previousMonthEnd)

        assertThat(analyzedCount).isEqualTo(0)
        verify(exactly = 0) { spendingAnalysisRepository.saveAnalysis(any()) }
    }
}
